### PR TITLE
[EDA-1390] TCL command for diagnostics

### PR DESCRIPTION
**Pull request by ravic-rs**
> ### Motivate of the pull request
> - [x] To address an existing issue. If so, please provide a link to the issue: EDA-1390
> - [ ] Breaking new feature. If so, please describe details in the description part.

> ### Describe the technical details
> #### What is currently done? (Provide issue link if applicable)
> New tcl command called "diagnostic" that will run a particular compiler task in diagnostic/debug mode

> #### What does this pull request change?
> > This PR adds diagnostic for packer only.

> ### Which part of the code base require a change
> 
> - [x] Library: <Specify the library name>
> - [ ] Plug-in: <Specify the plugin name>
> - [ ] Engine
> - [ ] Documentation
> - [ ] Regression tests
> - [ ] Continous Integration (CI) scripts

> ### Impact of the pull request

> - [ ] Require a change on Quality of Results (QoR)
> - [ ] Break back-compatibility. If so, please list who may be influenced.

### CODE REVIEW

File: src/Compiler/Compiler.cpp [hunk 1]
```diff
@@ -1754,6 +1759,21 @@ bool Compiler::RegisterCommands(TclInterpreter* interp, bool batchMode) {
   };
   interp->registerCmd("wave_refresh", wave_refresh, this, nullptr);
 
+  auto diagnostic = [](void* clientData, Tcl_Interp* interp, int argc,
```
Comment (alain-rs) [inline src/Compiler/Compiler.cpp:1762]: That diagnostic command should only set the option  (like your packer "debug"):
compiler->PackOpt(Compiler::PackingOpt::Debug)

The usage would be 
...
diagnostic "packer" ... eventually more options here
packing
...

**Review by alain-rs — CHANGES_REQUESTED** [review]

File: src/Compiler/Compiler.cpp (hunk 1 above)
Comment (KochynVolodymyr) [reply src/Compiler/Compiler.cpp:1762]: this is not what we agreed with @ravic-rs

Comment (ravic-rs) [reply src/Compiler/Compiler.cpp:1762]: I misinterpreted the requirement that diagnostic "packer" will run "packing" directly but not as a settings option before running the "packing"

**Review by alain-rs — COMMENTED** [review]
@ravic-rs,wWhere is the --device auto  substitution happening?

File: src/Compiler/Compiler.cpp (hunk 1 above)
Comment (ravic-rs) [reply src/Compiler/Compiler.cpp:1758]: We will rework this to support diagnostics for other tasks, for now this works.

**Comment by ravic-rs** [comment]
> @ravic-rs, where is the --device auto substitution happening for the 2nd run? I can't spot it. The device_size value should be substituted by "auto".

It's in foedag_rs: https://github.com/RapidSilicon/FOEDAG_rs/compare/main...task/EDA-1390/diagnostic-tcl-cmd

**Comment by alain-rs** [comment]
@KochynVolodymyr  to make complete sense for the open-source version the same change has to be also in Foedag
https://github.com/RapidSilicon/FOEDAG_rs/compare/main...task/EDA-1390/diagnostic-tcl-cmd

**Comment by ravic-rs** [comment]
> @KochynVolodymyr to make complete sense for the open-source version the same change has to be also in Foedag [RapidSilicon/FOEDAG_rs@main...task/EDA-1390/diagnostic-tcl-cmd](https://github.com/RapidSilicon/FOEDAG_rs/compare/main...task/EDA-1390/diagnostic-tcl-cmd)

Brought those changes here: https://github.com/os-fpga/FOEDAG/pull/1053/files#diff-d30b3f4529563e4e488f2f9ea11be3d3f651db9a28fd4d8ca5a83e314653206fR1703

**Pull request merged**